h/static/styles/core/_layout.scss

```diff
@@ -0,0 +1,167 @@
+// and internal columns. Each column has a gutter defined by $grid-gutter.
+// The implementation is largely based on the grid system created by
+// Harry Roberts <http://csswizardry.com/csswizardry-grids/> and the Yahoo
+// Pure system <http://git.io/ogODXA>
+
+// Sets up styles for the grid wrapper.
+@mixin grid-row($gutter: 0) {
+  margin: 0;
+  padding: 0;
+  // Remove the margin from the first column.
+  margin-left: -$gutter;
+  // Removes whitespace on browsers that do not support flexbox.
+  letter-spacing: -0.31em;
+  // Remove optimizeLegibility if applied.
+  text-rendering: optimizespeed;
+
+  // Removes whitespaec between elements in supporting browsers.
+  display: -webkit-flex;
+  -webkit-flex-flow: row wrap;
+  display: -ms-flexbox;
+  -ms-flex-flow: row wrap;
+
+  @content;
+}
+
+// Defines a column, can be included in any selector, widths can be provided
+// by passing a @content block.
+//
+// Example
+//
+// .my-awkward-item {
+//   @include grid-column($default-gutter) { width: 43%; }
+// }
+@mixin grid-column($gutter: 0) {
+  display: inline-block;
+  vertical-align: top;
+  padding-left: $gutter;
+  vertical-align: top;
+  width: 100%;
+  box-sizing: border-box;
+
+  // Reset letter spacing.
+  letter-spacing: normal;
+  text-rendering: auto;
+
+  @content;
+}
+
+// Defines selectors for a class based grid system. Only includes a few common
+// sizes at the moment, but can be expanded as necessary.
+@mixin grid-setup($namespace: "") {
+  .#{$namespace}1-1 { width: 100%; }
+  .#{$namespace}1-2 { width: 50%; }
+  .#{$namespace}1-3 { width: 33.333%; }
+  .#{$namespace}2-3 { width: 66.666%; }
+  .#{$namespace}1-4 { width: 25%; }
+  .#{$namespace}3-4 { width: 75%; }
+}
+
+
+@mixin breakpoint($min) {
+  @media only screen and (min-width: $min) {
+    @content;
+  }
+}
+
+// Mobile first media queries. Encourages development to work with mobile and
+// modify as the viewport grows rather than designing for individual bands.
+@mixin wide-handheld-and-up {
+  @include breakpoint($break-wide-handheld + 1) {
+    @content;
+  }
+}
+
+@mixin tablet-and-up {
+  @include breakpoint($break-tablet + 1) {
+    @content;
+  }
+}
+
+@mixin desktop-and-up {
+  @include breakpoint($break-desktop + 1) {
+    @content;
+  }
+}
+
+// Mixin for styling elements to make them more finger-friendly on touch-input
+// devices.
+//
+// Use interaction media queries where available (see
+// http://caniuse.com/#feat=css-media-interaction) or fall back to relying on a
+// JS-added class on the <html> or <body> elements otherwise.
+@mixin touch-input {
+  @media (pointer: coarse) {
+    @content;
+  }
+  .env-touch {
+    @content;
+  }
+}
+
+// Rules
+// ----------------------------------------------------------------------------
+
+// 1. Fixes for https://github.com/hypothesis/h/issues/4236
+// --------------------------------------------------------
+
+html {
+  display: flex; // 1
+}
```

Comment:
These are not generic rules but rather specific to the layout of the h homepage. Perhaps we should extract these as components, or just put them in the `site.scss` index file.

Reply:
Well, not all of the declarations there are for the fix mentioned in the comment, and all of these rules are currently applied in all three contexts (site, admin, and help-page).

I definitely want to avoid pulling this into a component. (Admittedly, this is partly because I'd like to kill "flexbox for global page layout" as soon as possible and therefore it'll hopefully only be around for a short while longer.) But the thinking is that eventually the `_// Rules_` section will contain classes that we can use to control the layout in the markup, so that we won't need much, if any, page-specific styling. Once we get there, "Rules" is where these kinds of styles (applicable to all layouts, like the ones here) would live.

Reply:
> Well, not all of the declarations there are for the fix mentioned in the comment, and all of these rules are currently applied in all three contexts (site, admin, and help-page).

Ah right, it was in `_base.scss` before. It shouldn't really have been there.

> Admittedly, this is partly because I'd like to kill "flexbox for global page layout" as soon as possible and therefore it'll hopefully only be around for a short while longer.

What do you want to do instead? You mentioned CSS grid before. If that's what you meant, we need to do some investigation that it is supported by a large enough proportion of our userbase before committing. ["Can I use" data](https://caniuse.com/#feat=css-grid) shows that it was only available generally in Chrome and Firefox as of March last year.

Reply:
I do want to use CSS grid, yes. It's the right tool for the job, and it's forward-looking.

You're correct that CSS grid only fully shipped early last year, but current support is at 85%, which makes it one of the fastest adoption curves for a CSS feature, ever. As you can see from the current CanIUse matrix, the only browsers that don't support it are Opera Mini and the Blackberry Browser (are these in our browser support matrix?); and QQ Browser and Baidu Browser (I assume not an issue until we tackle il8n generally). The desktop and mobile versions of IE 10/11 do only have partial support, but in their case that means that they support a previous version of the spec.

Having said all that, grid was fit for production even when it was shipped - this is Rachel Andrew, one of the authors of the spec, writing in April 2017: [https://rachelandrew.co.uk/archives/2016/01/29/vanishing-boxes-with-display-contents/](https://rachelandrew.co.uk/archives/2016/01/29/vanishing-boxes-with-display-contents/)

Reply:
> You're correct that CSS grid only fully shipped early last year, but current support is at 85%, which makes it one of the fastest adoption curves for a CSS feature, ever.

The number that would be super helpful for me to have is "what proportion of **our** visitors have browsers that support CSS grid well enough for us to use it"? If that is say 99% of our users then great, let's go for it. If on the other hand it is say 93% then we might need to be more careful.

We've got Google Analytics as one source of data. I can also pull user agent strings from our access logs. If there is a way that we can put a feature-testing canary in place (with `@supports`) that could also work. I'm open to other ideas.

> The desktop and mobile versions of IE 10/11 do only have partial support, but in their case that means that they support a previous version of the spec.

How different is that previous version? Just different property names and some edge-case behavioural differences or is it much more drastic than that? Will it be easy to handle with a postcss plugin (like we do for vendor prefixes) or is it more complex and something that developers will need to be mindful of?

Reply:
Would 93% support (and climbing) really not be enough? I think that's probably a bit too strict - [flexbox is only at 95%](https://caniuse.com/#feat=flexbox). Also, our current user base is important, but there's also our future users to consider as we grow, and the former may not be perfectly representative of the latter, thus the generic browser support stats are still useful.

Anyway, you should really read Rachel's article. I've posted it in a couple of places now, and it answers the questions you're asking. I'll pull two points from it: First, browsers that support grid will use it, and those that don't will ignore it, so graceful fallbacks are pretty easy. Second, there's no requirement for sites to look the same in every browser. Our site is responsive, so it already looks different on different devices and at different browser widths, to say nothing of users who consume our service with assistive technology. 

Also, let's not lose sight of the fact that IE 10/11 users are a vanishingly small percentage of our user base, and the internet in general. I'm attacking the bug reported by APA because they're a partner, and also because it's a useful wedge to talk about moving our codebase forward, but dying browsers should never guide our path.

Reply:
> Would 93% support (and climbing) really not be enough?

If hypothetically a feature was at 93% availability on _our_ site and it became a requirement, then we'd definitely have to be mindful of the fact that there are several thousand users coming to the site every month using browsers that don't support that feature. Exactly what happens for those users obviously depends on the feature and how important it is for the site.

> Second, there's no requirement for sites to look the same in every browser.

I agree. If the site doesn't look perfect on an old browser but the basic functionality still works, that may be acceptable in many cases.

> Also, let's not lose sight of the fact that IE 10/11 users are a vanishingly small percentage of our user base, and the internet in general. 

I looked at the data for our websites this morning and it turns out that IE 11 is actually not our biggest concern any more. See [this Slack thread](https://hypothes-is.slack.com/archives/C1M8NH76X/p1542358502081300).

> I'm attacking the bug reported by APA because they're a partner, and also because it's a useful wedge to talk about moving our codebase forward, but dying browsers should never guide our path.

I'm mostly concerned about being mindful of existing users who, for various reasons, are not using current browsers. That means at least doing some due diligence about the impact before making such a change. Fortunately in this case, [it looks like we're OK](https://hypothes-is.slack.com/archives/C1M8NH76X/p1542358502081300).